engine/execution/ingestion/throttle.go

```diff
@@ -0,0 +1,241 @@
+	c.log.Info().Msgf("initializing block throttle")
+	if c.inited {
+		return fmt.Errorf("throttle already inited")
+	}
+
+	c.inited = true
+	c.processables = processables
+
+	var unexecuted []flow.Identifier
+	var err error
+	if caughtUp(c.executed, c.finalized, c.threshold) {
+		unexecuted, err = findAllUnexecutedBlocks(c.state, c.headers, c.executed, c.finalized)
+		if err != nil {
+			return err
+		}
+		c.log.Info().Msgf("loaded %d unexecuted blocks", len(unexecuted))
+	} else {
+		unexecuted, err = findFinalized(c.state, c.headers, c.executed, c.executed+500)
+		if err != nil {
+			return err
+		}
+		c.log.Info().Msgf("loaded %d unexecuted finalized blocks", len(unexecuted))
+	}
+
+	for _, id := range unexecuted {
+		c.processables <- id
+	}
+
+	c.log.Info().Msgf("throttle initialized with %d unexecuted blocks", len(unexecuted))
+
+	return nil
+}
+
+func (c *BlockThrottle) OnBlockExecuted(_ flow.Identifier, executed uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.inited {
+		return fmt.Errorf("throttle not inited")
+	}
+
+	// we have already caught up, ignore
+	if c.caughtUp() {
+		return nil
+	}
+
+	// the execution is still far behind from finalization
+	c.executed = executed
+	if !c.caughtUp() {
+		return nil
+	}
+
+	c.log.Info().Uint64("executed", executed).Uint64("finalized", c.finalized).
+		Msgf("execution has caught up, processing remaining unexecuted blocks")
+
+	// if the execution have just caught up close enough to the latest finalized blocks,
+	// then process all unexecuted blocks, including finalized unexecuted and pending unexecuted
+	unexecuted, err := findAllUnexecutedBlocks(c.state, c.headers, c.executed, c.finalized)
+	if err != nil {
+		return fmt.Errorf("could not find unexecuted blocks for processing: %w", err)
+	}
+
+	c.log.Info().Int("unexecuted", len(unexecuted)).Msgf("forwarding unexecuted blocks")
+
+	for _, id := range unexecuted {
+		c.processables <- id
+	}
+
+	c.log.Info().Msgf("all unexecuted blocks have been processed")
+
+	return nil
+}
+
+func (c *BlockThrottle) OnBlock(blockID flow.Identifier) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.log.Debug().Msgf("recieved block (%v)", blockID)
+
+	if !c.inited {
+		return fmt.Errorf("throttle not inited")
+	}
+
+	// ignore the block if has not caught up.
+	if !c.caughtUp() {
+		return nil
+	}
+
+	// if has caught up, then process the block
+	c.processables <- blockID
+	c.log.Debug().Msgf("processed block (%v)", blockID)
+
+	return nil
+}
+
+func (c *BlockThrottle) OnBlockFinalized(lastFinalized *flow.Header) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.inited {
+		return
+	}
+
+	if c.caughtUp() {
+		return
+	}
+
+	if lastFinalized.Height <= c.finalized {
+		return
+	}
+
+	c.finalized = lastFinalized.Height
+}
+
+func (c *BlockThrottle) caughtUp() bool {
+	return caughtUp(c.executed, c.finalized, c.threshold)
+}
+
+func caughtUp(executed, finalized uint64, threshold int) bool {
+	return finalized <= executed+uint64(threshold)
+}
```

Comment:
It is so that easier to write test cases for this pure function, otherwise I would have to create the `BlockThrottle` module for testing this condition.